File: src/main/java/leetcode/Problem69.java
package leetcode;

public class Problem69 {
    public int mySqrt(int x) {
        if (x < 2) {
            return x;
        }
        return (int) mySqrt(x, 2, x/2);
    }

    private long mySqrt(long x, long lo, long hi) {
        if (lo > hi) {
            return hi;
        }
        long mid = (lo + hi) / 2;
        if (mid * mid > x) {
            return mySqrt(x, lo, mid - 1);
        }
        return mySqrt(x, mid + 1, hi);
    }
}

File: src/problem_69.py
class Solution:
    def mySqrt(self, x: int) -> int:
        lo, hi = 0, x
        while lo < hi:
            mi = hi - (hi - lo) // 2
            if mi * mi > x:
                hi = mi - 1
            else:
                lo = mi
        return lo
